fix 2d array bug of abiSchemaToJsonSchema in web3-validator

EtlesL opened the pull request:
## Description
We encountered a bug in the web3-validator code where the handling of multi-dimensional arrays was not implemented correctly.
In the original code snippet, the arraySize vatiable was only assigned the value of the first element in the arraySizes array. This prevented proper handling of multi-dimensional arrays.

```typescript
const arraySize = arraySizes[0];
const item: JsonSchema = {
    type: 'array',
    $id: abiName,
    items: convertEthType(String(baseType)),
    minItems: arraySize,
    maxItems: arraySize,
};
```

To fix the bug, I added an additional loop to iterate over the arraySizes array to handle each dimension:
```typescript
for (let i = 0; i < arraySizes.length; i++) {
  const arraySize = arraySizes[i];
  const item: JsonSchema = {
      type: 'array',
      $id: abiName,
      items: convertEthType(String(baseType)),
      minItems: arraySize,
      maxItems: arraySize,
  };

  if (arraySize < 0) {
      delete item.maxItems;
      delete item.minItems;
  }

  (lastSchema.items as JsonSchema[]).push(item);
}
```

Besides, different from Issue #6486, if the multi-dimensional arrays has a fixed length, it will go to different code snippet.

Fixes #6798 

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

## Checklist:

- [x] I have selected the correct base branch.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [ ] Any dependent changes have been merged and published in downstream modules.
- [x] I ran `npm run lint` with success and extended the tests and types if necessary.
- [x] I ran `npm run test:unit` with success.
- [x] I ran `npm run test:coverage` and my test cases cover all the lines and branches of the added code.
- [ ] I ran `npm run build` and tested `dist/web3.min.js` in a browser.
- [x] I have tested my code on the live network.
- [ ] I have checked the Deploy Preview and it looks correct.
- [x] I have updated the `CHANGELOG.md` file in the root folder.
- [x] I have linked Issue(s) with this PR in "Linked Issues" menu.

Commit:
fix 2d array bug of abiSchemaToJsonSchema in web3-validator

Commit:
fix 2d array bug of abiSchemaToJsonSchema in web3-validator(add comments)

jdevcs commented on packages/web3-validator/src/utils.ts:
```diff
@@ -205,37 +205,43 @@ export const abiSchemaToJsonSchema = (
-
-			(lastSchema.items as JsonSchema[]).push(item);
+		    // iterate over arraySizes array to each dimension
+		    for (let i = 0; i < arraySizes.length; i++) {
```
@EtlesL 
some of fixes are required for Linter to pass https://github.com/web3/web3.js/actions/runs/8058948881/job/22029093796?pr=6836#step:5:246 , also could you add unit tests for this usecase, Thanks for your contribution,

Commit:
fix abi json schema test case:
- 123 cannot be uint
- test case("nested tuple object in nested array") has a wrong data structure of tuple[][3]

Commit:
fix function abiSchemaToJsonSchema in iterating over arraySizes array to each dimension

Commit:
FIX:
1. fix function convertToZod to distinguish tuple and array, add checks in maxItems & minItems
2. revert abiSchemaToJsonSchema of utils and fix
3. add testcases in validator.validate

EtlesL commented:
@jdevcs 
I'm sorry for the misunderstanding about the type conversion earlier. 
At first, I thought the multi-dimensional array was not converted correctly, so I added a loop. However, after carefully reading the code, I realized that this was not correct. Regarding this bug,
```typescript
public validate(schema: JsonSchema, data: Json, options?: { silent?: boolean }) {
    const zod = convertToZod(schema);
    const result = zod.safeParse(data);
    if (!result.success) {
	    const errors = this.convertErrors(result.error?.issues ?? []);
	    if (errors) {
		    if (options?.silent) {
			    return errors;
		    }
		    throw new Web3ValidatorError(errors);
	    }
    }
    return undefined;
}
```
The core issue lies in the first line, where the schema is converted to Zod. In convertToZod function, multi-dimensional arrays and tuple types are confused.
```typescript
if (schema?.type === 'array' && schema?.items) {
    if (Array.isArray(schema.items) && schema.items.length > 0) {
        // do somsthing
        return z.tuple(arr as [ZodTypeAny, ...ZodTypeAny[]]);
    }
    return z.array(convertToZod(schema.items));
}
```
So, I added additional conditions to distinguish between them. If it's a tuple
1. schema.items MUST be an array
2. the length schema.items MUST have more than 1 items
3. maxItems MUST NOT be undefined
4. the ids of `schema.items` MUST be unique
```typescript
if (Array.isArray(schema.items) 
    && schema.items.length > 1
    && schema.maxItems !== undefined
    && new Set(schema.items.map((item: JsonSchema) => item.$id)).size === schema.items.length) {
        //do something
        return z.tuple(arr as [ZodTypeAny, ...ZodTypeAny[]]);
}
```
Also, in this method, the maxItems and minItems were lost during the conversion, resulting in the length judgment always being "must NOT have more than 1 item."
```typescript
const nextSchema = Array.isArray(schema.items) ? schema.items[0] : schema.items;
let zodArraySchema = z.array(convertToZod(nextSchema));

zodArraySchema = schema.minItems !== undefined ? zodArraySchema.min(schema.minItems) : zodArraySchema;
zodArraySchema = schema.maxItems !== undefined ? zodArraySchema.max(schema.maxItems) : zodArraySchema;
return zodArraySchema;
```

Besides that, the previous unit tests did not cover all test cases. After adding the abiToJsonSchemaCases in the web_validator.test, I found that:
1. The test case "nested tuple object in nested array" has a type error. The tuple[][3] should have 3 elements in the outer layer and an unspecified number in the inner layer, but the original code was written in reverse.
2. The test case "nested tuple object in nested array" also has a small issue with the utils abi to jsonSchema, which I fixed as well.

Please review again, thx.

Commit:
UPDATE: contract test

EtlesL committed:
Merge branch '4.x' into 4.x

EtlesL commented:
I can't run tests of web-eth-contracts on my local environment, I saw the error in the auto pipeline and fixed. Not sure why, because I didn't change the codes of contracts packages. Please review.

EtlesL commented:
Hi, are there any updates? I noticed the pipeline failed, but there are no error logs provided.

jdevcs commented:
Thanks for detailed explanation of changes, there are some unstable tests ( not related with this PR ) and we are working in current sprint for fixing these. https://github.com/web3/web3.js/pull/6857, could you rebase your branch when 6857 is merged,

EtlesL committed:
Merge branch '4.x' into 4.x

EtlesL commented:
> Thanks for detailed explanation of changes, there are some unstable tests ( not related with this PR ) and we are working in current sprint for fixing these. #6857, could you rebase your branch when 6857 is merged,

@jdevcs Hi, I saw it merged, and I've updated the branch.

EtlesL committed:
Merge branch '4.x' into 4.x
